apps/nextjs/src/stores/lessonPlanStore/stateActionFunctions/handleMessagesUpdated.ts

```diff
@@ -0,0 +1,122 @@
+import {
+  applyLessonPlanPatchImmutable,
+  extractPatches,
+  type PatchDocument,
+} from "@oakai/aila/src/protocol/jsonPatchProtocol";
+import { LessonPlanKeySchema } from "@oakai/aila/src/protocol/schema";
+import { aiLogger } from "@oakai/logger";
+import { createHash } from "crypto";
+
+import type { AiMessage } from "@/stores/chatStore/types";
+
+import type { LessonPlanStore } from "..";
+
+const log = aiLogger("lessons:store");
+
+const timeOperation = <T>(fn: () => T): T => {
+  const startTime = performance.now();
+  const result = fn();
+  const endTime = performance.now();
+  log.info(`applyLessonPlanPatch took ${endTime - startTime} milliseconds`);
+  return result;
+};
+
+const generatePatchHash = (patch: PatchDocument): string => {
+  const patchString = JSON.stringify(patch);
+  const hash = createHash("sha256");
+  hash.update(patchString);
+  return hash.digest("hex");
+};
+
+const countCompleteParts = (messageStr: string) => {
+  return [...messageStr.matchAll(/"status":"complete"/g)].length;
+};
```

Comment:
A stringy function to extract the number of patches which have been streamed to completion